lenskit-core/src/main/java/org/lenskit/data/dao/file/DelimitedColumnEntityFormat.java

```diff
@@ -0,0 +1,235 @@
+            throw new RuntimeException("could not instantiate entity builder", e);
+        }
+    }
+
+    /**
+     * Add a column.
+     * @param attr The attribute to add as a column, or `null` to skip the next column.
+     * @throws IllegalStateException if non-labeled columns have already been specified.
+     */
+    public void addColumn(@Nullable TypedName<?> attr) {
+        if (columns == null) {
+            Preconditions.checkState(labeledColumns == null, "mixed labeled and unlabeled columns");
+            columns = new ArrayList<>();
+        }
+        columns.add(attr);
+    }
+
+    /**
+     * Add columns to a format.
+     * @param columns The columns to add.
+     */
+    public void addColumns(TypedName<?>... columns) {
+        for (TypedName<?> col: columns) {
+            addColumn(col);
+        }
+    }
+
+    /**
+     * Add a column.
+     * @param label The header label.
+     * @param attr The attribute to add as a column, or `null` to skip the next column.
+     * @throws IllegalStateException if non-labeled columns have already been specified.
+     */
+    public void addColumn(String label, @Nullable TypedName<?> attr) {
+        if (labeledColumns == null) {
+            Preconditions.checkState(columns == null, "mixed labeled and unlabeled columns");
+            labeledColumns = new LinkedHashMap<>();
+        }
+        labeledColumns.put(label, attr);
+    }
+
+    @Override
+    public LineEntityParser makeParser(List<String> header) {
+        assert header.size() == getHeaderLines();
+
+        if (usesHeader() && labeledColumns != null) {
+            assert header.size() == 1;
+            List<TypedName<?>> cols = new ArrayList<>();
+            StrTokenizer tok = new StrTokenizer(header.get(0), delimiter);
+            while (tok.hasNext()) {
+                String label = tok.next();
+                cols.add(labeledColumns.get(label));
+            }
+            return new OrderedParser(cols, tok);
+        } else {
+            return new OrderedParser(columns, new StrTokenizer("", delimiter));
+        }
+    }
+
+    private class OrderedParser extends LineEntityParser {
+        int lineNo = 0;
+        StrTokenizer tokenizer;
+        List<TypedName<?>> fileColumns;
+
+        public OrderedParser(List<TypedName<?>> columns, StrTokenizer tok) {
+            fileColumns = columns;
+            tokenizer = tok;
+        }
+
+        @Override
+        public Entity parse(String line) {
+            tokenizer.reset(line);
+            lineNo += 1;
+
+            EntityBuilder builder = newEntityBuilder()
+                    .setId(lineNo);
+
+            // since ID is already set, a subsequent ID column will properly override
+
+            for (TypedName column: fileColumns) {
+                String value = tokenizer.nextToken();
+                if (value != null && column != null) {
+                    builder.setAttribute(column, column.parseString(value));
```

Comment:
Yes. A particular entity builder, such as `RatingBuilder`, is free to fail when it gets to `build()` — that is how I intend to enforce mandatory fields, rather than making field specification more complicated.